### PR TITLE
[Relax][Utils] Add ExtendFunc utils

### PR DESCRIPTION
**Pull request by SiriusNEO**
**Background:** suppose we have a backbone module `mod` which has a function `forward`. Run `forward` we can get the output/predictions of this module. Now to run AD in this module, we need to "extend" the function `forward` with a specified loss function. This can be easily done if we have an Inline pass. We can define a new function
```
def forward_loss(*args, labels):
    out = forward(*args)
    loss = loss_func(out, labels)
    return loss
``` 
and inline it. But implementing an Inline is difficult and it needs call graph analysis (which still lacks in relax). So for now we can use this simple util to replace Inline here temporarily.
This util is similar with "tail call inline". The following is an example:
```
# Before.
@R.function
def func1(a, b):
    return a + b, a * b

@R.function
def func2(c, d, e):
    return d, c, c + e

# After. func1_func2 = extend_func(orig_func=func1, ex_func=func2).
@R.function
def func1_func2(a, b, e):
    c = a + b
    d = a * b
    return d, c, c + e
```

### CODE REVIEW

**Comment by MasterJH5574** [comment]
My major concern for this PR mainly comes from the specification on the function boundary.

By placing the function under `relax/util.h`, this PR is meant to work for generic settings, which you may face the pair of functions like
```python
## Example 1
@R.function
def func_first(t: R.Tuple([R.Tensor(), R.Tensor()])) -> R.Tuple([R.Tensor(), R.Tensor()]):
  return t

@R.function
def func_second(t: R.Tuple([R.Tensor(), R.Tensor()])):
  ...


## Example 2
@R.function
def func_first(t: R.Tuple([R.Tensor(), R.Tensor()])) -> R.Tuple([R.Tensor(), R.Tensor()]):
  return t

@R.function
def func_second(x: R.Tensor(), y: R.Tensor()):
  ...
```
It is not straightforward to explain the specification clearly and increases the complexity to do jobs well pretty much.

Another concern is about the behavior of lifting additional param of the second func to the fused function’s param list. It worths thinking whether this behavior is often desired, or people expect to pose a rejection to the inputs.

I understand that this util is mainly for loss function use at this moment. It is very nice for you to think about possible generalization opportunities. But for this specific case, I think the generalization the util to various input functions is trickier than simply inline the call to loss functions, where you clearly know that the callee is a loss function and have a clear picture of the function boundary between the caller and callee.

So I might suggest leaving this as a local simple util for training (specifically, loss function) at this moment, which is much less complex. Simplicity is not a bad thing.

**Comment by SiriusNEO** [comment]
It will be refactored and I will bring it together with loss functions in this PR: https://github.com/mlc-ai/relax/pull/112. Let's close it.
